### ISSUE TITLE
Add support for * and ? wildcards

### ISSUE THREAD

**Issue by theflakes** [jopohl/sigma-rust#8]
Hello, and thank you for this crate.

I'm playing with it and can't get the below to work. I am seeing the rule loaded correctly as well as the log. The problem seems to be the substring search. I can match the full string.

rule
```
id: e781ae02-0d7b-4c3a-b1e3-d3b6b24f6c5b
title: Example Sigma Rule
description: Detects example log entries
status: experimental
author: Brian Kellogg
tags:
  - test
logsource:
  product: windows
  service: sysmon
detection:
  error:
    norm_keys.messsage: '*error*'
  condition: error
falsepositives:
  - NA
level: high
```

Log
```
{"count":5,"enabled":"true","lvl":"error","msg":"This is an error message","norm_keys":{"enabled":true,"level":"error","message":"This is an error message"},"sigma_rules":[],"tags":["example","test"]}
```

Sigma match function
```
fn apply_sigma_rules(log_entry: &Value, rules: &Vec<Rule>) -> Vec<Value> {
    let mut matched_rules = Vec::new();
    let event = event_from_json(&log_entry.to_string()).unwrap();

    for rule in rules {
        if rule.is_match(&event) {
            matched_rules.push(Value::String(rule.title.clone()));
        } else {
            println!("Rule did not match!");
        }
    }

    matched_rules
}
```

**Comment by jopohl** [comment]
Hi! To clarify, you expect the condition `norm_keys.messsage: '*error*'` to match `... "norm_keys":{"enabled":true,"level":"error","message":"This is an error message"} ...`? On a side note, in  `norm_keys.messsage` the word "message" is written with three "s". 

I think you rather want the `contains` modifier in this case because otherwise the engine performs an exact string matching for `*error*`, i.e., your message would need to be something like `"This is an *error* message"`.

To sum up, I suggest to rewrite your rule as follows.

```yaml
id: e781ae02-0d7b-4c3a-b1e3-d3b6b24f6c5b
title: Example Sigma Rule
description: Detects example log entries
status: experimental
author: Brian Kellogg
tags:
  - test
logsource:
  product: windows
  service: sysmon
detection:
  error:
    norm_keys.message|contains: 'error'
  condition: error
falsepositives:
  - NA
level: high
```

Does this work for you?

**Comment by theflakes** [comment]
Sorry, mistyped message obviously on the above issue, the rule I'm using locally is correct though. I believe the spec for Sigma includes two wildcards in simple string searches; '?', and '*'. See detection section: https://github.com/SigmaHQ/sigma-specification/blob/main/specification/sigma-rules-specification.md

![image](https://github.com/user-attachments/assets/c759b23c-393d-46f3-a641-659665890132)

**Comment by jopohl** [comment]
Thank you, good link. I used the sigmahq website (https://sigmahq.io/docs/basics/rules.html) when implementing this library as a reference but apparently on their GitHub they provide further specifications.

**Comment by theflakes** [comment]
Maybe this will work in the `values.rs` source file? Just using glob for contains and changing `startswith` and `endswith` to regex.

```
   pub(crate) fn contains(&self, other: &Self) -> bool {
        match (self, other) {
            (Self::String(a), Self::String(b)) => glob_match(b, a),
            _ => false,
        }
    }

    pub(crate) fn starts_with(&self, other: &Self) -> bool {
        match (self, other) {
            (Self::String(a), Self::String(b)) => {
                let mut pattern = String::new();
                let mut chars = b.chars().peekable();
    
                while let Some(ch) = chars.next() {
                    match ch {
                        '\\' => {
                            if let Some(next_ch) = chars.peek() {
                                pattern.push('\\');
                                pattern.push(*next_ch);
                                chars.next();
                            }
                        },
                        '*' => pattern.push_str(".*"),
                        '?' => pattern.push('.'),
                        _ => pattern.push_str(&regex::escape(&ch.to_string())),
                    }
                }
    
                let pattern = format!("^{}", &pattern);
                let re = Regex::new(&pattern).unwrap();
                re.is_match(a)
            },
            _ => false,
        }
    }
    
   
    pub(crate) fn ends_with(&self, other: &Self) -> bool {
        match (self, other) {
            (Self::String(a), Self::String(b)) => {
                let mut pattern = String::new();
                let mut chars = b.chars().peekable();
    
                while let Some(ch) = chars.next() {
                    match ch {
                        '\\' => {
                            if let Some(next_ch) = chars.peek() {
                                pattern.push('\\');
                                pattern.push(*next_ch);
                                chars.next();
                            }
                        },
                        '*' => pattern.push_str(".*"),
                        '?' => pattern.push('.'),
                        _ => pattern.push_str(&regex::escape(&ch.to_string())),
                    }
                }
    
                let pattern = format!("{}$", &pattern);
                let re = Regex::new(&pattern).unwrap();
                re.is_match(a)
            },
            _ => false,
        }
    }
```

**Comment by theflakes** [comment]
Man, I've been playing with this and not at all satisfied with my solutions. Lose a lot of EPS by scanning strings for globs and accounting for if they escaped. I've tried a couple crates with no luck so far; wildcard, wildflower.

**Comment by theflakes** [comment]
Thanks, I've got simple glob matching implemented here: https://github.com/theflakes/sigma-rust

**Comment by theflakes** [comment]
I'm really interested to see how you code this and compare it against how much I over thought my solution.

**Comment by jopohl** [comment]
Thanks for bringing up this issue; was fun to implement a wildcard matching algorithm from scratch.
